File: coding/src/com/algos/li14_linkedlist/_L83_RemoveDuplicatesFromSortedList.java
package edu.algos.li14_linkedlist;

import edu.algos.li30_model.ListNode;

public class _L83_RemoveDuplicatesFromSortedList {

	public static void main(String[] args) {
		_L83_RemoveDuplicatesFromSortedList obj = new _L83_RemoveDuplicatesFromSortedList();
		
		_G01_SingleLinkedList list = new _G01_SingleLinkedList();
		list.insertAtEnd(2);
		list.insertAtEnd(2);
		list.insertAtEnd(3);
		list.insertAtEnd(3);
		list.insertAtEnd(4);
		
		list.head = obj.deleteDuplicates(list.head);
		list.print();
	}
	
	// Problem: https://leetcode.com/problems/remove-duplicates-from-sorted-list/
	// Idea: Compare curr and curr.next vals, if equal, change curr.next to curr.next.next.
	public ListNode deleteDuplicates(ListNode head) {
        ListNode curr = head;
        while(curr != null && curr.next != null){
            if(curr.val == curr.next.val){
                curr.next = curr.next.next;
            }else{
                curr = curr.next;
            }
        }
        
        return head;
    }

}

File: dsa/topics/09_linkedlist/python/l83_remove_duplicates_from_sorted_list.py
"""
Problem: https://leetcode.com/problems/remove-duplicates-from-sorted-list/
Idea:
Compare curr and curr.next vals.
If equal, change curr.next to curr.next.next.
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def delete_duplicates(head: ListNode) -> ListNode:
    curr = head

    while curr and curr.next:
        if curr.val == curr.next.val:
            curr.next = curr.next.next
        else:
            curr = curr.next

    return head


# Example usage
if __name__ == "__main__":
    head = ListNode(2,
            ListNode(2,
            ListNode(3,
            ListNode(3,
            ListNode(4)))))

    head = delete_duplicates(head)

    while head:
        print(head.val, end=" ")
        head = head.next
